### ISSUE TITLE
Sublime Syntax compatibility

### ISSUE THREAD

**Issue by borela** [tree-sitter/tree-sitter#139]
Sublime's syntax is a way more powerful than the tmLanguage and allows you to control the contexts (an example of JavaScript expressions https://github.com/borela/naomi/blob/master/syntaxes/fjsx15/expression.sublime-syntax), is possible to somehow convert the syntax into the tree sitter grammar?

**Comment by maxbrunsfeld** [comment]
No, I don't think it's possible. The differences between the two systems are much deeper than syntax. 

Tree-sitter grammars are [context free grammars](https://en.wikipedia.org/wiki/Context-free_grammar) which completely describe the structure of the language. They are used to create parsers, which produce exact syntax trees with a similar level of precision as parsers that are used in compilers and interpreters.

Sublime text grammars may have more feature than TextMate grammars, but they are similar in purpose. They allow you to approximately match certain constructs in a language using regexes and a stack. They aren't based on a sound underlying theory that would allow you to efficiently *parse* source code.

**Comment by borela** [comment]
> They allow you to approximately match certain constructs in a language using regexes and a stack. They aren't based on a sound underlying theory that would allow you to efficiently parse source code.

Not true, you can give the correct meaning to tokens using sublime syntax, the only current limitation is that the regex engine limits patterns per line which could be fixed:

https://github.com/SublimeTextIssues/Core/issues/1693

Another benefit is that the YAML syntax is alot clearer.

**Comment by maxbrunsfeld** [comment]
> Not true, you can give the correct meaning to tokens using sublime syntax.

I'm not exactly sure what exactly you mean by 'give the correct meaning to tokens', but that address what I'm saying. Sublime grammars *are not designed to produce syntax trees*. They are designed to recognize *some* language constructs in a source file by running regexes over the source file.

Does that make sense? You seem to be familiar with the sublime grammar system. Let me try to give you an example of how Tree-sitter is different. Consider this javascript code:

```js
async function test(name, {a, b}) {
  for (const c of a) {
    await b.d(
      c,
      <div>
        Hello {name}
      </div>,
      (f) => f.g
    );
  }
}
```

Here is the output of parsing that with [tree-sitter-javascript](https://github.com/tree-sitter/tree-sitter-javascript):

```
(program [0, 0] - [11, 0]
  (function [0, 0] - [10, 1]
    (identifier [0, 15] - [0, 19])
    (formal_parameters [0, 19] - [0, 33]
      (identifier [0, 20] - [0, 24])
      (object_pattern [0, 26] - [0, 32]
        (shorthand_property_identifier [0, 27] - [0, 28])
        (shorthand_property_identifier [0, 30] - [0, 31])))
    (statement_block [0, 34] - [10, 1]
      (for_of_statement [1, 2] - [9, 3]
        (identifier [1, 13] - [1, 14])
        (identifier [1, 18] - [1, 19])
        (statement_block [1, 21] - [9, 3]
          (expression_statement [2, 4] - [8, 6]
            (await_expression [2, 4] - [8, 5]
              (call_expression [2, 10] - [8, 5]
                (member_expression [2, 10] - [2, 13]
                  (identifier [2, 10] - [2, 11])
                  (property_identifier [2, 12] - [2, 13]))
                (arguments [2, 13] - [8, 5]
                  (identifier [3, 6] - [3, 7])
                  (jsx_element [4, 6] - [6, 12]
                    (jsx_opening_element [4, 6] - [4, 11]
                      (identifier [4, 7] - [4, 10]))
                    (jsx_text [4, 11] - [5, 14])
                    (jsx_expression [5, 14] - [5, 20]
                      (identifier [5, 15] - [5, 19]))
                    (jsx_text [5, 20] - [6, 6])
                    (jsx_closing_element [6, 6] - [6, 12]
                      (identifier [6, 8] - [6, 11])))
                  (arrow_function [7, 6] - [7, 16]
                    (formal_parameters [7, 6] - [7, 9]
                      (identifier [7, 7] - [7, 8]))
                    (member_expression [7, 13] - [7, 16]
                      (identifier [7, 13] - [7, 14])
                      (property_identifier [7, 15] - [7, 16]))))))))))))
```

Every single token in the tree is interpreted and fits precisely into a well-defined tree structure. This is similar what you would get from parsing the file with a standard JavaScript parsing tool like [esprima](http://esprima.org/).

Do you understand how Esprima is different than sublime grammars? There is a big difference, and Tree-sitter's functionality is a lot more like Esprima than sublime.

**Comment by maxbrunsfeld** [comment]
> Another benefit is that the YAML syntax is alot clearer.

I'm assuming you haven't looked at a Tree-sitter grammar yet, even though I responded to this in #138. Here's the [JavaScript grammar](https://github.com/tree-sitter/tree-sitter-javascript/blob/master/grammar.js). It is much, much easier to read than anything you could express this content in YAML. It's not even close.

**Comment by borela** [comment]
> Here is the output of parsing that with tree-sitter-javascript:

I see, so it's a combination of a tokenizer and a parser, can tree sitter generate the tree for partially valid code or would it stop at the moment it finds a bad token? 

As for trees you can infer them from the scopes assigned by the syntax in the case of sublime syntax.

**Comment by borela** [comment]
I'll try to build the project and look at the docs, thanks.

**Comment by maxbrunsfeld** [comment]
> can tree sitter generate the tree for partially valid code or would it stop at the moment it finds a bad token?

Yeah, a major unique part of Tree-sitter is an error recovery algorithm that allows it to generate a syntax tree for partially valid code. For example, in the above code, if I accidentally changed `async` on line 1 to `asyn`, tree-sitter would produce the same tree, except there would be an `ERROR` node at the beginning:

```
(program [0, 0] - [11, 0]
  (ERROR [0, 0] - [0, 4]
    (identifier [0, 0] - [0, 4]))
  (function [0, 5] - [10, 1]
    (identifier [0, 14] - [0, 18])
    (formal_parameters [0, 18] - [0, 32]
      (identifier [0, 19] - [0, 23])
      (object_pattern [0, 25] - [0, 31]
        (shorthand_property_identifier [0, 26] - [0, 27])
        (shorthand_property_identifier [0, 29] - [0, 30])))
    (statement_block [0, 33] - [10, 1]
      (for_of_statement [1, 2] - [9, 3]
        (identifier [1, 13] - [1, 14])
        (identifier [1, 18] - [1, 19])
        (statement_block [1, 21] - [9, 3]
          (expression_statement [2, 4] - [8, 6]
            (await_expression [2, 4] - [8, 5]
              (call_expression [2, 10] - [8, 5]
                (member_expression [2, 10] - [2, 13]
                  (identifier [2, 10] - [2, 11])
                  (property_identifier [2, 12] - [2, 13]))
                (arguments [2, 13] - [8, 5]
                  (identifier [3, 6] - [3, 7])
                  (jsx_element [4, 6] - [6, 12]
                    (jsx_opening_element [4, 6] - [4, 11]
                      (identifier [4, 7] - [4, 10]))
                    (jsx_text [4, 11] - [5, 14])
                    (jsx_expression [5, 14] - [5, 20]
                      (identifier [5, 15] - [5, 19]))
                    (jsx_text [5, 20] - [6, 6])
                    (jsx_closing_element [6, 6] - [6, 12]
                      (identifier [6, 8] - [6, 11])))
                  (arrow_function [7, 6] - [7, 16]
                    (formal_parameters [7, 6] - [7, 9]
                      (identifier [7, 7] - [7, 8]))
                    (member_expression [7, 13] - [7, 16]
                      (identifier [7, 13] - [7, 14])
                      (property_identifier [7, 15] - [7, 16]))))))))))))

```

> As for trees you can infer them from the scopes assigned by the syntax in the case of sublime syntax.

Yeah, this is true. But I don't think you would get as precise of a syntax tree. I'd be curious what inferred tree sublime would give you for the above code.

**Comment by maxbrunsfeld** [comment]
Here are the docs: http://tree-sitter.github.io/tree-sitter. They are very incomplete right now 😢 , but I plan to work on them more soon.

**Comment by borela** [comment]
> For example, in the above code, if I accidentally changed async on line 1 to asyn, tree-sitter would produce the same tree

That's impressive! I didn't know this was possible.

> Here are the docs: http://tree-sitter.github.io/tree-sitter. They are very incomplete right now 😢 , but I plan to work on them more soon.

No problem, it looks good for now, I'll try to create a grammar for MQL5 and if I find something odd I'll bother you again lol.

**Comment by Thom1729** [comment]
@maxbrunsfeld Sublime syntax definitions are essentially specifications for nondeterministic pushdown automata, and they can recognize nondeterministic context-free languages.

The lack of nondeterminism is relevant to some languages (see [proposal](https://github.com/SublimeTextIssues/Core/issues/2241) for examples), but overall Sublime's engine is very similar in expressive power to tree-sitter. It is misleading to characterize it as more similar to TextMate's engine than to tree-sitter; a deterministic context-free language is much more similar to a nondeterministic context-free language than it is to a regular language.

> They allow you to approximately match certain constructs in a language using regexes and a stack. They aren't based on a sound underlying theory that would allow you to efficiently parse source code.

This is misleading in particular, because “regexes and a stack” encompass pushdown automata, which are indeed “a sound underlying theory that would allow you to efficiently parse source code”. The fact that tree-sitter syntaxes are defined via grammars rather than via PDAs is immaterial.

(In fact, Sublime's support for non-regular regexps that use backreferences allows it to parse some constructs that are not even context-free, although this is a fairly minor point. I am curious, though: how does tree-sitter handle heredocs?)

I'm not trying to start an editor war; tree-sitter is a fine system, and as a formal language nerd I love to see the lesser-known algorithms used to solve practical problems. But I've run into more than one person who came away from tree-sitter presentations with significant misconceptions about how Sublime's engine works.

**Comment by maxbrunsfeld** [comment]
I’m not sure what you mean when you say that Sublime’s paradigm is closer to Tree-sitter than to Text-mates. With both Textmate and Sublime, you have a stack in addition to the regexes.

Fundamentally, in both cases, you *manually* write out all the states of a push-down automaton. Emphasis on the word *manually*! I don’t think it’s misleading to characterize this as a far less powerful parsing paradigm than the one Tree-sitter uses.

It’s true that in *theory*, you can write a sublime/textmate grammar that’s as powerful as an ordinary LR parser. After all, an LR parser is just a push-down automaton that’s generated based on a declarative grammar, instead of being written down by hand. In *practice* though, the quality is very very different. A generated LR parser for, say, Python, as you might find in a compiler, would have a few thousand different states. That’s what’s necessary to really model the language token by token, building a full syntax tree.

Despite being very long in terms of line count (much much longer than Tree-sitter grammars), sublime grammars usually only have about a hundred states. Transitions between these states are pretty coarse grained; the regexes often match several tokens at a time, assuming that they’ll appear on the same line. In some languages, it’s adequate for mostly-correct syntax highlighting, but it’s really not an appropriate tool for building a full parse tree. The automatons required for that are just too complex; you don’t want to write them down by hand, state by state. And the fact is, people don’t. Which is why sublime and textmate grammars to this day don’t do a satisfying job of highlighting very popular languages, especially ones with types, like C, C++, Go, Java, C#, and Rust, where it’s really nice to see the types highlighted consistently.

And that’s not even getting into issues of ambiguity and lookahead that Tree-sitter solves by using GLR. These issues come up for lots of languages, and you can only *approximately* solve them with sublime/textmate grammars (AFAIK) by using long, complex regexes with multiple tokens of lookahead, which is bad for performance and makes all kinds of assumptions about layout.

To summarize: yes, there really is a theoretical difference in recognizing power between GLR and a push-down automaton. But even beyond the theoretical difference, there is a very large practical difference as well, because of the line-by-line nature of Textmate and Sublime grammars and the incredible tedium of writing a DFA for an entire programming language by hand.

**Comment by Thom1729** [comment]
You're talking about two different things, neither of which is relevant to the objection I made.

First, you say that writing a tree-sitter grammar is easier than writing a Sublime syntax definition. That's a matter of opinion, but it's certainly not an unreasonable opinion. I actually like tree-sitter; I have nothing against it. I can't comment on the relative ease of use, though, because I have a great deal more experience with one than with the other; it wouldn't be a fair comparison. In either case, as you concede, this difference in ease of use does not prevent Sublime from parsing nondeterministic context-free languages. It's fairly obvious that the sublime-syntax format was deliberately designed to have that level of power (unless its author just accidentally invented one of the best-known types of automaton by accident).

Second, you point out that many (or even most) Sublime syntax definitions resemble old TextMate definitions. This is almost trivially true, because most Sublime syntax definitions in the wild were converted from TextMate definitions. There are automated tools for doing exactly this, and these tools are widely used by the Sublime community. The reason for this is that a) sublime-syntax is essentially a superset of tmLanguage, b) it's much easier to work with, and c) it runs a lot faster in Sublime. So if you have an old TextMate definition, and you're not interested in rewriting it, then it's still worth it to convert it to a sublime-syntax.

The reason that this isn't true of tree-sitter is that there is no migration path. Correct me if I'm wrong, but as far as I know there is no way to automatically convert a tmLanguage to a tree-sitter grammar. If there were, then I have no doubt that you would see practically overnight a flood of tree-sitter grammars that were basically fancied-up tmLanguages.

So when you compare the existing body of tree-sitter grammars to sublime-syntaxes, you're comparing a handful of brand-new grammars written by yourself and other early adopters to an enormous corpus of syntax definitions mostly written by newer users, a large fraction of which are simply tmLanguages that were auto-converted to take advantage of Sublime's optimized regexp engine.

A useful case study is Sublime's JavaScript syntax. Before sublime-syntax, it was a tmLanguage, of course. When the new engine became available three years ago, the old syntax was replaced with a direct sublime-syntax conversion of [JavaScript Next](https://github.com/Benvie/JavaScriptNext.tmLanguage). Since then, it has been almost totally rewritten to take full advantage of the new features. If you take a look at the [current version](https://github.com/sublimehq/Packages/blob/master/JavaScript/JavaScript.sublime-syntax), you'll see that it's parsing the code at a deep and detailed level. The sole limitation — which I've mentioned before — is that Sublime's engine does not implement nondeterminism.

Rewriting the JavaScript syntax was a slow process because I didn't want to break any of the tests. At present, the JavaScript syntax tests include about six thousand assertions. I could have rewritten the whole thing from scratch in significantly less time, but this would have unnecessarily disrupted highlighting for existing users.

Some specific points I'd like to address:

> A generated LR parser for, say, Python

Python is not context-free. Neither an LR nor a GLR parser can correctly parse Python.

> especially ones with types, like C, C++, Go, Java, C#, and Rust, where it’s really nice to see the types highlighted consistently.

This is an aesthetic choice, not an engine limitation. It's a common convention to highlight the same identifier differently in different contexts (such as declared/used). If you don't like this, you can probably change your color scheme to highlight it the way you want.

Except, of course, that a CFG can't tell in `a * b;` whether `a` is a type. This is fundamentally a context-sensitive problem. Real C parsers handle this by building up a symbol table during lexical analysis. I checked the tree-sitter documentation and I did not see any indication that this was possible. If it is, then you've been drastically underselling the formal power of tree-sitter!

> And that’s not even getting into issues of ambiguity and lookahead that Tree-sitter solves by using GLR. These issues come up for lots of languages, and you can only approximately solve them with sublime/textmate grammars (AFAIK) by using long, complex regexes with multiple tokens of lookahead … To summarize: yes, there really is a theoretical difference in recognizing power between GLR and a push-down automaton.

Here you're referring (I presume) to nondeterminism. To a first approximation, Sublime syntaxes handle this via cover grammars. (Interestingly, this is actually how arrow function syntax is defined [in the spec](https://www.ecma-international.org/ecma-262/#prod-CoverParenthesizedExpressionAndArrowParameterList).) While this prevents the highlighting from breaking due to wrong guesses, it does mean that in some cases the arguments won't be highlighted correctly. This is, undoubtedly, an advantage for tree-sitter, but it's important to note that it's already known how to augment sublime-syntax to parse nondeterministically; there just hasn't been the demand to justify the implementation.

> …which is bad for performance…

Here you're referring to long chains of regexp captures that are typical of old TextMate syntaxes. In more modern Sublime syntaxes, captures are not needed and are used much more sparingly. This does have a small but measurable effect on performance. Overall, though, I would expect Sublime's engine to perform much faster than tree-sitter. I'm not sure how to run benchmarks in Atom, but if you could refer me to the appropriate command I'll post some results.

> …the incredible tedium of writing a DFA for an entire programming language by hand.

DPDA. :-)

> Despite being very long in terms of line count (much much longer than Tree-sitter grammars)…

Looking in the grammar repos has left me a bit confused. I checked the C grammar to try and figure out how it was (or wasn't) disambiguating types from values in statements like `a * b;`, and in the `src` directory I found ~4,500 lines of JSON and ~60,000 lines of C. Is this actually compiled from the ~800-line `grammar.js` file in the root?

I also checked the Ruby grammar to see how heredocs are implemented (suspecting that the more rigid structure of tree-sitter would preclude Sublime's solution) and it seems like the problem is solved via ~900 lines of C code, and it looks like `scanner.cc` was handwritten, not compiled from something simpler. What's the deal here?

**Comment by Thom1729** [comment]
While I'm thinking of it, one important point of clarification: Sublime's parser isn't technically context-free. In fact, because it supports arbitrary Oniguruma regexps, it can parse some constructs that are not context-free. For instance, you could use it to match arbitrary XML with missing tags. This is how Sublime syntaxes implement heredocs. Contexts that use backreferences or some other Oniguruma features trigger the old Oniguruma regexp engine, but other contexts in the same syntax will still use the new, faster engine.

This capability isn't widely used (because it's rarely useful for parsing real languages), but it does exist. I mention it because it seems to me that tree-sitter should be restricted to true-regular regexps, which I presume is why tree-sitter-ruby seems to implement heredocs with handwritten C.

**Comment by maxbrunsfeld** [comment]
Yeah, things like heredocs in ruby and indent/dedent tokens in python are implemented in Tree-sitter via an external scanner: a bit of arbitrary C/C++/Rust code that handles a specific set of tokens. This is the only way to really match the behavior of these constructs in their native parsers. As described [here](https://github.com/sublimehq/Packages/issues/430#issuecomment-234606995), Sublime does not attempt to handle heredocs in general (e.g. multiple heredocs starting on one line). It also doesn’t attempt to handle Python’s fundamental indent/dedent structure either, right? Fancy regex features aren’t really sufficient for handling these.

**Comment by maxbrunsfeld** [comment]
My intention isn’t to disparage Sublime’s syntax highlighting in particular. And I definitely don’t want to mislead anyone, which is why I’ve never attempted to explain in detail how other systems besides Tree-sitter works.

What I *have* said is that Tree-sitter is significantly different than systems like Textmates and Sublime’s regarding how accurately it attempts to parse languages. I absolutely stand by this, and I think it’s an important thing for people to understand, when explaining what Tree-sitter is.

Let me put it this way: What’s the most accurate sublime text parser for, say, Python, Ruby, C, C++, JavaScript, Typescript, Go, or Rust? I’d love to take it for a spin and find out that it recognizes all of the same stuff as a Tree-sitter parser. Just like I’d be willing to try out a Firefox clone written in x86 assembly, if it were really better than Firefox.

**Comment by Thom1729** [comment]
> Yeah, things like heredocs in ruby and indent/dedent tokens in python are implemented in Tree-sitter via an external scanner: a bit of arbitrary C/C++/Rust code that handles a specific set of tokens. This is the only way to really match the behavior of these constructs in their native parsers.

True, but the native parsers do a lot more than highlight the code.

> As described here, Sublime does not attempt to handle heredocs in general (e.g. multiple heredocs starting on one line).

That seems to be specific to Ruby's heredocs, which have some pretty wild syntax. Heredocs in most languages aren't nearly so troublesome to parse. Honestly, I'd like to know how Ruby's parser handles it. You'd think that if they were implementing it the hard way in the scanner that it would have occurred to them that it might be a bad idea. I say all of this as a Ruby fan.

> It also doesn’t attempt to handle Python’s fundamental indent/dedent structure either, right?

No. Is this ever necessary for syntax highlighting?

> What’s the most accurate sublime text parser for, say, Python, Ruby, C, C++, JavaScript, Typescript, Go, or Rust?

For JavaScript, the core syntax should be good. (There's a bit of a PR backlog, so there are some bugs that aren't fixed in master.) If you need JSX, Flow, or somesuch, then [JS Custom](https://github.com/Thom1729/Sublime-JS-Custom) will dynamically recompile a syntax definition with your choice of optional features. Unfortunately, I haven't yet implemented TypeScript as an optional extra.

Using only basic functionality (no backrefs or exotic regexps), the JavaScript syntax:

- Handles the automatic semicolon insertion algorithm in all known cases.
- Correctly distinguishes comma operators from other syntactic uses of the comma.
- Never confuses division with regexp literals even in pathological cases.

JS Custom uses mostly the same code, but also shows off Sublime's ability to embed syntaxes arbitrarily (via the `custom_template_tags` option).

I mention these things in particular (and I mention the JavaScript syntax in particular) because these are notoriously difficult for simple, TextMate-style regexp scanners to figure out. In order to handle these situations reliably, the syntax has to do the same kind of work as any "real" JavaScript parser. The implementation ignores some irrelevant details like operator precedence, but if you examine the source you'll see how many of the contexts map directly onto productions in the ECMA 262 formal grammar. (After the core of the overhaul was completed, we received bug reports from users who complained that the syntax was broken only to discover that Sublime was right and their code didn't do what they thought it did. In at least one case, a user reported a parsing bug and we discovered that both Babel and Uglify had the same bug.)

All that said, there are a couple of known issues that can, in fact, be attributed to limitations of the Sublime engine:

- Reliably identifying arrow functions requires nondeterminism, which Sublime does not provide. Instead, we use heuristics. If we guess wrong, the highlighting won't break disruptively, but formal parameter names may not be highlighted as such. Exact highlighting will resume when the `=>` is encountered. Because tree-sitter supports nondeterminism, I expect that it would handle arrow functions exactly.
- `yield` and `await` are always treated as keywords, even outside generator/async functions. The relevant productions in the ECMA grammar are parameterized, a notational solecism disguising the fact that there are essentially four slightly different versions of each production. We could implement this exactly, but it would require a very large amount of near-duplicate code. In principle, pathological code could break highlighting, but I doubt very much that any human being will ever write such a thing by accident. From a glance at the tree-sitter-javascript syntax, I expect that it would suffer from the same bug.
- In `import.meta`, `import` will be the wrong color if it is followed by a line break.

I won't say that there are no other bugs, but if there are then I expect them to be fixable within Sublime's engine and within the structure that the syntax currently uses.

If you want a more apples-to-apples comparison, I can come up with a slimmed-down version of the JavaScript syntax that does all of the same "real" work but omits legacy behavior like special treatment of `$` in identifiers. (That kind of crud takes up a surprising amount of space, unfortunately. I'm hoping eventually to transition that sort of thing to configurable options.)

---

With that in mind, I would characterize Sublime's engine and tree-sitter as fundamentally similar systems that nevertheless differ in several ways:

- The two systems have very different approaches to syntax definition: automaton-based versus grammar-based. Sublime's are more "low-level".
- Tree-sitter can handle nondeterministic languages, which is useful for some constructs in some languages.
- Sublime supports full Oniguruma regexps, which lets it handle some context-sensitive constructs natively.
- Tree-sitter can't handle context-sensitive constructs at the grammar level, but you can write a custom lexer in C that can handle even more cases than Sublime.
- Sublime's engine is probably faster. (I'm prepared to be wrong about this.)

In addition, I have some questions that I think would help clarify my understanding of tree-sitter and its relationship to Sublime:

- Looking at the tree-sitter-javascript custom scanner, it seems like the scanner itself has special code to handle embedded expressions in template strings. Is this necessary, or just convenient? More generally, how does one tree-sitter syntax embed another?
- The same scanner contains code to handle automatic semicolon insertion. Is this just a convenience, or is there some obstacle to implementing this directly in the parser?

**Comment by mitranim** [comment]
While I have no opinion on the technical merits of tree-sitter vs. sublime-syntax, I'm concerned about potential misunderstanding and misrepresentation. Max seems to conflate what sublime-syntax _can_ do with what many syntax authors _chose_ to do. I'd like to clear that up with an example.

The November talk introducing tree-sitter had a Go example. The tree-sitter definition detected types where they're expected to appear, and seemed to make no distinction between built-in or user-defined types. The Sublime definition highlighted built-in types and ignored user-defined types, which the presenter found undesirable (and so do I). The presenter appeared to blame sublime-syntax. At that time, Sublime's Packages repository already had an open [PR](https://github.com/sublimehq/Packages/pull/1662) for a better Go definition, which has since been merged and ships with the dev build. The new syntax parses the language like the compiler does. Among other benefits, it detects types contextually, without relying on built-ins. This was done in sublime-syntax; the old definition simply wasn't _trying_ to solve this problem.

By misplacing the blame on sublime-syntax, tree-sitter was presented as a solution to a non-problem, a strawman.

I'm not saying Atom and GH should have adopted sublime-syntax instead. Subjective arguments about ease of use and brevity are perfectly valid, and I'm not equipped to compare technical merits. I just want tree-sitter/Atom/GH folks to realize that the actual power of sublime-syntax cannot be evaluated by looking at most existing syntaxes. One needs to pick from a small sample of recent definitions explicitly written to match a language's AST, such as JSCustom, recent Go, recent Clojure. By the way, this answers @maxbrunsfeld's question about which "most accurate parser" to look at. (The latter two are incubating in https://github.com/sublimehq/Packages `master` and ship with ST's dev build. I'm happy to instruct how to get that into stable.)

**Comment by maxbrunsfeld** [comment]
> Max seems to conflate what sublime-syntax can do with what many syntax authors chose to do.

> One needs to pick from a small sample of recent definitions explicitly written to match a language's AST, such as JSCustom, recent Go, recent Clojure.

Yeah, @Thom1729 said something similar to this earlier in the thread.

Here's my assessment of what I said in the talk:

* I said that Sublime Text did not correctly highlight types in Go. This was absolutely true at the time.
* I said that Sublime Text didn't recognize things like types in Go because of a fundamental limitation. **This actually is not true, in the specific case of Go**.
* Sublime Text has a fundamentally less powerful parsing system than Tree-sitter. **This is absolutely true**.

If I were giving the talk today, I would not use the example of types in Go
* I *still* might use the example of types in other languages, like TypeScript and Rust.
* To describe the fundamental improvements that Tree-sitter provides over Sublime, I would use different examples.

---

> By misplacing the blame on sublime-syntax, tree-sitter was presented as a solution to a non-problem, a strawman.

I don't understand how you could think this, unless you only watched the part of the presentation that talked about Sublime Text, which was just a couple of sentences.

Tree-sitter is not just a syntax highlighting system! The thesis behind Tree-sitter is that incremental parsing can allow us to really *parse* code in a text editor. We can then perform syntax highlighting based on the concrete syntax trees, and we can also use the same syntax trees for many other kinds of useful code analysis, even outside of text editors. The parsers that Tree-sitter produces are *still* very different from the `JSCustom` or `Go` Sublime grammars. 

Those Sublime grammars are not parsers. Despite being significantly longer and more complex than the corresponding Tree-sitter grammars, they produce *much* less information. A few examples (some of which we already discussed above):

* > The implementation ignores some irrelevant details like operator precedence

    Operator precedence is absolutely irrelevant for syntax highlighting, but if you're trying to parse code (as Tree-sitter grammars are) it is a major component of the job.

* > `yield` and `await` are always treated as keywords. ...  From a glance at the tree-sitter-javascript syntax, I expect that it would suffer from the same bug.
   
    No, Tree-sitter handles functions and variables named `yield` and `await` correctly. One limitation is that it will *permit* await expressions outside of async functions, but in general I am ok with producing a valid parse for code that is technically invalid.

* > identifying arrow functions requires nondeterminism

    As you mention, Tree-sitter can parse arrow functions correctly. Furthermore, if you are trying to actually parse code, this type of problem is pretty widespread.

* Sublime Text still as no way to handle things like indentation in Python, which is fundamental to Python's grammar. Tree-sitter has a system for handling arbitrary imperative features like this, while still maintaining incremental parsing.

* Sublime Text still has no way of handling things like Ruby's Heredocs in a general way, as we discussed above.

**Comment by maxbrunsfeld** [comment]
@Thom1729 I forgot to respond to some things that you said earlier, since I was on vacation during that convo.

> DPDA. :-)

A push-down automata is basically a finite-state-automata plus a stack of states. When discussing the state-machine *piece* in particular, I think it's reasonably clear to refer to that *piece* using the common term "DFA". I unde

> Python is not context-free. Neither an LR nor a GLR parser can correctly parse Python.

It actually *is* generally recognized that Python is described by a context-free grammar. Specifically, by this *official* grammar available on python.org: https://docs.python.org/3/reference/grammar.html.  The part that I assume you're calling "not context-free", the indentation rules, is regarded there as separate from the grammar itself. Behold, an LR parser that correctly parses python: [tree-sitter-python](https://github.com/tree-sitter/tree-sitter-python)! 

The larger point that I was trying to make though (in which Python was just an arbitrary example) is that fundamentally, you need on the order of 2000 states in your state machine, in order to parse that language. So in Sublime's grammar system, that means hand-writing 2000 interlinked YAML objects.

> Except, of course, that a CFG can't tell in `a * b`; whether `a` is a type. This is fundamentally a context-sensitive problem. Real C parsers handle this by building up a symbol table during lexical analysis. I checked the tree-sitter documentation and I did not see any indication that this was possible. 

Indeed, the C family is especially tricky. Not exactly for the reason that you cite though. The problem of `a * b` is not as bad as you make it sound; using GLR (what you've referred to as non-determinism), you can try *both* interpretations, and in the majority of cases, only one interpretation will be valid. In the remaining cases, one case will be much more likely than the other, which can be expressed in the grammar using *dynamic precedence*.

<details>
<summary>Example Output</summary>

```sh
$ cat test.c
```

```c
int main() {
  // multiplication or pointer declaration?
  a * b;      // pointer declaration
  c * d == 5; // multiplication
  
  // call or parenthesized declaration?
  e (f);    // call
  int (h);  // parenthesized declaration
}
```

```sh
$ tree-sitter parse test.c
```

```clj
(translation_unit [0, 0] - [9, 0]
  (function_definition [0, 0] - [8, 1]
    (primitive_type [0, 0] - [0, 3])
    (function_declarator [0, 4] - [0, 10]
      (identifier [0, 4] - [0, 8])
      (parameter_list [0, 8] - [0, 10]))
    (compound_statement [0, 11] - [8, 1]
      (comment [1, 2] - [1, 43])
      (declaration [2, 2] - [2, 8]
        (type_identifier [2, 2] - [2, 3])
        (pointer_declarator [2, 4] - [2, 7]
          (identifier [2, 6] - [2, 7])))
      (comment [2, 14] - [2, 36])
      (expression_statement [3, 2] - [3, 13]
        (equality_expression [3, 2] - [3, 12]
          (math_expression [3, 2] - [3, 7]
            (identifier [3, 2] - [3, 3])
            (identifier [3, 6] - [3, 7]))
          (number_literal [3, 11] - [3, 12])))
      (comment [3, 14] - [3, 31])
      (comment [5, 2] - [5, 39])
      (expression_statement [6, 2] - [6, 8]
        (call_expression [6, 2] - [6, 7]
          (identifier [6, 2] - [6, 3])
          (argument_list [6, 4] - [6, 7]
            (identifier [6, 5] - [6, 6]))))
      (comment [6, 12] - [6, 19])
      (declaration [7, 2] - [7, 10]
        (primitive_type [7, 2] - [7, 5])
        (identifier [7, 7] - [7, 8]))
      (comment [7, 12] - [7, 40]))))
```

</details>

<br>


> Overall, though, I would expect Sublime's engine to perform much faster than tree-sitter.

I'm not certain, but I doubt it. On [the `syntect` README](https://github.com/trishume/syntect#performance) they report that Sublime takes 98ms to parse jQuery 2.1, which is *much* slower than parsing it with Tree-sitter on my laptop. I'd like to compare them directly on my machine, but I don't know how to benchmark Sublime parsers. Benchmarking Tree-sitter parsers is very easy though; you can just run `tree-sitter parse <the-file> --time` at the command line.

**Comment by maxbrunsfeld** [comment]
> The same scanner contains code to handle automatic semicolon insertion. Is this just a convenience, or is there some obstacle to implementing this directly in the parser?

I actually used to handle it directly in the grammar, but I was finding that it was a bit less efficient, because it caused ambiguity in some cases. Does your sublime grammar handle this correctly?

```js
foo()

// ↑ this is a function call, but there's no automatic semicolon,
// because the token *after* the function call is a `[`.

// this is another comment.

// ↓ this is a subscript into the return value of `foo()`

[1] = 2;
```

**Comment by maxbrunsfeld** [comment]
@Thom1729 @mitranim Overall, your point is well taken that Sublime Text has a more parsing system than I gave it credit for in my brief statements about Sublime during my StrangeLoop talk. I can see how it would be frustrating to hear my statement, given that you two have personally invested a lot of effort into Sublime grammars, and Sublime's grammar ecosystem *is* starting to improve.

On the other hand, I still think that this original issue - Sublime Syntax compatibility, is not a feature that makes sense for Tree-sitter, because Sublime grammars are so fundamentally different from Tree-sitter grammars. I also still think that Tree-sitter offers meaningful value that Sublime's grammar system does not provide. You both may still disagree about that, and if so, I must give up now on trying to persuade you. The best thing for us all to do is probably to continue working on our own separate solutions to this problem.

I need to lock this issue at this point, because I check GitHub notifications throughout my work day, and discussions like these are too time-consuming and emotionally draining for me to keep entering into them. I'm not trying to imply that either of you made unprofessional comments; just need to lock it for my own sanity. Feel free to tweet at me if you have other thoughts.
